pkg/rebuild/rebuild/muddlestrategy_test.go

```diff
@@ -0,0 +1,413 @@
+		{
+			name: "full_config",
+			strategy: MuddleStrategy{
+				Location: Location{
+					Dir:  "test-dir",
+					Repo: "https://example.com/test-repo",
+					Ref:  "abc123",
+				},
+				Source: []MuddleStep{
+					{Runs: "echo source"},
+					{Uses: "git-checkout"},
+				},
+				Deps: []MuddleStep{
+					{
+						Uses: "npm/install",
+						With: map[string]string{
+							"npmVersion": "8.0.0",
+						},
+					},
+				},
+				Build: []MuddleStep{
+					{Runs: "make build"},
+				},
+				SystemDeps: []string{"git", "npm"},
+				OutputPath: "dist/output",
+			},
+			wantYAML: `
+location:
+    repo: https://example.com/test-repo
+    ref: abc123
+    dir: test-dir
+src:
+    - runs: echo source
+    - uses: git-checkout
+deps:
+    - uses: npm/install
+      with:
+        npmVersion: 8.0.0
+build:
+    - runs: make build
+system_deps:
+    - git
+    - npm
+output_path: dist/output
+`,
+		},
+		{
+			name: "minimal_config",
+			strategy: MuddleStrategy{
+				Location: Location{
+					Repo: "https://example.com/test-repo",
+					Ref:  "abc123",
+				},
+			},
+			wantYAML: `
+location:
+    repo: https://example.com/test-repo
+    ref: abc123
+`,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Marshal the struct to YAML
+			gotYAML, err := yaml.Marshal(tc.strategy)
+			if err != nil {
+				t.Fatalf("yaml.Marshal() error = %v", err)
+			}
+
+			// Compare generated YAML with expected YAML (normalizing whitespace)
+			if diff := cmp.Diff(strings.TrimSpace(tc.wantYAML), strings.TrimSpace(string(gotYAML))); diff != "" {
+				t.Errorf("YAML mismatch (-want +got):\n%s", diff)
+			}
+
+			// Unmarshal back to struct
+			var gotStrategy MuddleStrategy
+			if err := yaml.Unmarshal(gotYAML, &gotStrategy); err != nil {
+				t.Fatalf("yaml.Unmarshal() error = %v", err)
+			}
+
+			// Compare original struct with round-tripped struct
+			if diff := cmp.Diff(tc.strategy, gotStrategy); diff != "" {
+				t.Errorf("Round-trip mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestBuiltinCommand_GitCheckout(t *testing.T) {
+	tests := []struct {
+		name     string
+		location Location
+		buildEnv BuildEnv
+		want     string
+	}{
+		{
+			name: "fresh_checkout",
+			location: Location{
+				Repo: "https://github.com/test/repo",
+				Ref:  "deadbeef",
+			},
+			buildEnv: BuildEnv{HasRepo: false},
+			want:     "git clone https://github.com/test/repo .\ngit checkout --force 'deadbeef'",
+		},
+		{
+			name: "existing_repo",
+			location: Location{
+				Repo: "https://github.com/test/repo",
+				Ref:  "0000",
+			},
+			buildEnv: BuildEnv{HasRepo: true},
+			want:     "git checkout --force '0000'",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cmd := builtinCommands["git-checkout"]
+			if cmd == nil {
+				t.Fatal("git-checkout command not found")
+			}
+
+			data := struct {
+				With     map[string]string
+				Target   Target
+				BuildEnv BuildEnv
+				Location Location
+			}{
```

Comment:
Should this use the same typedef as the system under test? I'm leaning yes because we would want this test to break if the builtin is depending on a field that's not present in the real data struct.

Reply:
Actually, why not use generateForStep in this test?

Reply:
good point. this was added a few refactors before that existed.